### ISSUE TITLE
Settings/Config not retained, or radomly retained, even in the config/save files.

### ISSUE THREAD

**Issue by Casteele72** [freeorion/freeorion#3386]
Operating System:
$ uname -all
Linux earth 5.4.0-0.bpo.4-amd64 #1 SMP Debian 5.4.19-1~bpo10+1 (2020-03-09) x86_64 GNU/Linux

FreeOrion Version:
0.4.8 [build 2018-08-23.26f16b0] CMake (installed from the standard Debian repository)

Issue:
The settings/config has a mind of it's own?!
When starting a new single player game, I tweak a number of settings, and then start the game. If I decide to resign the game and start a new one, various setting exhibit any or all of the following:
-- The settings go back to their default;
-- The settings are retained in the settings display, but do not remain when starting a new game;
-- The settings display the defaults, but retain their previous values when a new game starts.
Thus far, the only way I have been able to ensure all settings are to my liking is to go through and edit each one: Add one then subtract one for numeric values, add and then delete a space after any text, or toggle options on then back off (or off then back on) prior to starting the game.

It also seems to randomly include/exclude settings in the config.xml and persistent_config.xml files. Sometimes it will include/exclude items that are the same as the defaults, and other times it will include/exclude items that are different than the defaults.

In all cases, I am not able to reproduce the exact same results as the results are unpredictable and random each time. However, I would suggest that when writing the config file(s), it writes out all settings--even the ones identical to the defaults. If nothing else, it will give clues on what the setting/config parts of the code "see" as the current settings/config.

Another side note: At times, it also does not always seem to honor settings in the config files, current settings, or in save games. As an example of the above issues, I tried to start a game with the Super Tester Takeover option and saved the game. The option was saved in the save file the first game, but not in the second one I started without changing any options in between. (I would provide the save files, except that they were auto-saves and have since been overwritten with newer auto-saves.)

**Comment by geoffthemedio** [comment]
Did you really test 0.4.8?

That release is 2.5 years old and the relevant code has likely changed quite a bit in newer versions, so the reported behaviour is likely at least partially different now...

**Comment by agrrr3** [comment]
> Did you really test 0.4.8?
> 
That is the version in debian stable (buster). So probably yes.

 @Casteele72 the current stable version is 0.4.10, i am not sure if that builds with libraries in buster or not. We usually do bugfixes only for the master branch (the upcoming release, works well, but is unbalanced and frequently breaks save games). If you simply want to play on buster, easiest is probably to install the flatpak/snap (which include the dependencies which work). The master branch is also provided as a weekly build on the latest/beta snapcraft channel.

@geoffthemedio @vezzra is such an issue (if it persists) to warant a bugfix release of 0.4.10?

**Comment by geoffthemedio** [comment]
probably not, but if Vezzra wants to do another bugfix release, sure...

**Comment by Casteele72** [comment]
I have grabbed the v0.4.10.1 release branch from GIT, and will test to see if the problems persist. OTOH, since the version I played is the main version available to the Debian 10 users (which may not have the desire, knowledge, or resources to compile the latest from source), perhaps someone with more knowledge of the source code can work with the Debian package maintainer to get it either updated, or at least a bug fix for the distributed version? (I would try, but that is entering an area I'm not really very familiar with...)

**Comment by Casteele72** [comment]
So far, preliminary testing results show that most of the problems at the main menu level persist in v0.4.10.1. The exact version string from my build is:
v0.4.10.1 HEAD [build 2020-09-25.39cfe10] CMake

At the level of running/playing a game, it seems to save/restore all the settings I have tested thus far--Although when I try to load my save games from the previous version used, the galaxy comes up empty and there are error messages in he chat/message window. (I will not post them here--If the developers want me to file a bug report about that, I will file a separate bug report with all relevant files. Let me know.)

**Comment by geoffthemedio** [comment]
> when I try to load my save games from the previous

New versions are generally not expected to be compatible with old saves.

Most of the setup options in the single player galaxy setup are not expected to be remembered between runs of the program. But if you have cases where the setup options show one setting when starting a game but something else is set in-game in the pedia game rules listing, that is a bug.

**Comment by Casteele72** [comment]
Re-saves: That is too bad; Perhaps a notice alerting users to that should be displayed. However, that is an unrelated issue here, so I will leave that be or open a suggestion ticket for it.

Re-settings: Please clarify "between runs"? Especially with how the game uses multiple binaries for the "Main Program/Menu", "Clients" and "AIs", and the "Server/Daemon". If I start a single player game, and at some point I decide I do not like the tweaks I made to the settings, I go back to the main menu (labelled "Resign" in my previous version) to tweak them some more. I am not completely restarting FreeOrion, just going back to the main menu to adjust and start a new game with a little more tweaking.

So perhaps it is not a coding bug, but I would still classify it as a "logic bug", because as a user I expect the galaxy settings to remain where I last set them. Even between runs. This is especially true because FreeOrion has many settings to customize game play (I have been trying to find a good balance between size, memory, CPU, and game length--It is frustrating trying to remember all the settings I tweaked) and is certainly one of the better OSI remakes that has an overall polished look and feel that many others lack. I also have the additional thought that so few OSI games really inspire users to learn to code or join the code teams; I would like to see FO change that.

None the less, if it would make it easier, feel free to reclassify this as a "feature request" if that makes more sense. For right now, it seems to be saving (and displaying) the correct settings in save games, so that is no longer a major issue or priority.

**Comment by geoffthemedio** [comment]
> Re-saves: That is too bad; Perhaps a notice alerting users to that should be displayed. However, that is an unrelated issue here, so I will leave that be or open a suggestion ticket for it.

To elaborate a bit, there actually is compatibility code that is maintained during development of each version, but usually by the time the next point release is ready, so much has change that it doesn't even make sense to try to load an old save.

> Re-settings: Please clarify "between runs"?

Shutting down the client "freeorion" program and restarting it is a new "run" in this context.

> If I start a single player game, and at some point I decide I do not like the tweaks I made to the settings, I go back to the main menu (labelled "Resign" in my previous version) to tweak them some more. I am not completely restarting FreeOrion, just going back to the main menu to adjust and start a new game with a little more tweaking.

If you resign (or Exit to Menu in the newer versions), it should remember the galaxy settings (on the left) from before, but not the game rules within the tabs to the right. This is a missing feature.

It should also remember the galaxy settings if you exit the program and run it again. At least it does for me.

**Comment by Casteele72** [comment]
Yes. The galaxy size and such is saved. The game rules are not. It is the game rules that for me, I have to do he most adjusting to find a good game setup. Mostly trying to find a good balance with things like research and production rates, ship strength and speed, and so on. I personally feel this is one of the better qualities of the game: With the right settings, I can have the AI or other human players starting to expand in to my area in under 100 turns, or more than a couple hundred. When playing with other humans, they tend to like faster pace games that only last an evening. When playing alone, I like the feel of the slow exploration and expansion in to the unknown, not knowing when I will encounter a hostile empire.

Perhaps eventually it will be possible to specify different config file names to store different settings for different scenarios like that. (I might even provide patch sets for that, as I learn more about the code structure...)

**Comment by hifron** [comment]
Why there is no specie/race properties which could be researched or somehow gained?

**Comment by Vezzra** [comment]
@agrrr3,

> is such an issue (if it persists) to warant a bugfix release of 0.4.10?

As it turned out, the only issue that remained in 0.4.10 (if I understand correctly), is that game rule settings aren't stored in `config.xml` yet. As @geoffthemedio pointed out, that's a missing feature.

As bugfix releases are only intended to provide bug fixes for rather serious/severe bugs that somehow slipped through, and not to backport missing features to the release branch, a bugfix release for this is unlikely. Unless there is some very good reason to do it...
